src/primitives/iter.rs

```diff
@@ -0,0 +1,501 @@
+/// checks for this.
+#[derive(Clone, PartialEq, Eq)]
+pub struct FeToByteIter<I: Iterator<Item = Fe32>> {
+    last_fe: Option<Fe32>,
+    bit_offset: usize,
+    iter: I,
+}
+
+impl<I> Iterator for FeToByteIter<I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    type Item = u8;
+    fn next(&mut self) -> Option<u8> {
+        let bit_offset = {
+            let ret = self.bit_offset;
+            self.bit_offset = (self.bit_offset + 8) % 5;
+            ret
+        };
+
+        if let Some(last) = self.last_fe {
+            let mut ret = last.0 << (3 + bit_offset);
+
+            self.last_fe = self.iter.next();
+            let next1 = self.last_fe?;
+            if bit_offset > 2 {
+                self.last_fe = self.iter.next();
+                let next2 = self.last_fe?;
+                ret |= next1.0 << (bit_offset - 2);
+                ret |= next2.0 >> (7 - bit_offset);
+            } else {
+                ret |= next1.0 >> (2 - bit_offset);
+                if self.bit_offset == 0 {
+                    self.last_fe = self.iter.next();
+                }
+            }
+
+            Some(ret)
+        } else {
+            None
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (fes_min, fes_max) = self.iter.size_hint();
+        let min = fes_len_to_bytes_len(fes_min);
+        let max = fes_max.map(|max| fes_len_to_bytes_len(max) + 1);
+        (min, max)
+    }
+}
+
+/// Iterator adaptor that just prepends a single character to a field element stream.
+///
+/// More ergonomic to use than `std::iter::once(fe).chain(iter)`.
+pub struct WitnessVersionIter<I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    witness_version: Option<Fe32>,
+    iter: I,
+}
+
+impl<I> Iterator for WitnessVersionIter<I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    type Item = Fe32;
+
+    fn next(&mut self) -> Option<Fe32> { self.witness_version.take().or_else(|| self.iter.next()) }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let hint = self.iter.size_hint();
+        match self.witness_version {
+            None => hint,
+            Some(_) => (hint.0, hint.1.map(|max| max + 1)),
+        }
+    }
+}
+
+/// Iterator adaptor for field-element-yielding iterator, which tacks a
+/// checksum onto the end of the yielded data.
+#[derive(Clone, PartialEq, Eq)]
+pub struct ChecksumIter<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    iter: I,
+    checksum_remaining: usize,
+    checksum_engine: checksum::Engine<Ck>,
+}
+
+impl<I, Ck> ChecksumIter<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    /// Helper function to input an HRP into the underlying checksum engine of the iterator.
+    ///
+    /// This function is infallible, but if you feed it a non-ASCII `hrp` it probably
+    /// will cause your checksum engine to produce useless results.
+    pub fn with_checksummed_hrp(mut self, hrp: &Hrp) -> Self {
```

Comment:
In 0c6613106fa624db037598dab5e9a19634acd2d0:

Maybe in the doccomment also mention that "if you use this after the iterator has already yielded elements" that this probably will also produce useless results.

Reply:
Done!